File: src/app/shell_integration.zig
//! Shell integration setup: write the embedded zsh/bash scripts to a runtime
//! dir and export the env vars that wire spawned shells to them.
//!
//! `Pty.buildChildEnv` copies `environ` into every child, so env vars exported
//! here (before any tab spawns) are inherited by every shell.

const std = @import("std");

// setenv/unsetenv are standard POSIX but not declared in Zig 0.16's std.c.
extern "c" fn setenv(name: [*:0]const u8, value: [*:0]const u8, overwrite: c_int) c_int;
extern "c" fn unsetenv(name: [*:0]const u8) c_int;

const integration_zsh = @embedFile("../shell/caldera-integration.zsh");
const integration_bash = @embedFile("../shell/caldera-integration.bash");
const zdotdir_zshenv = @embedFile("../shell/zdotdir-zshenv.zsh");

/// Resolve `~/.cache/caldera-console/shell` into `buf`. Null when `$HOME`
/// is unset.
fn runtimeDir(buf: []u8) ?[]const u8 {
    const home = std.c.getenv("HOME") orelse return null;
    const h = std.mem.span(home);
    return std.fmt.bufPrint(buf, "{s}/.cache/caldera-console/shell", .{h}) catch null;
}

/// Create every directory along `path` (like `mkdir -p`). Best-effort.
fn mkdirP(path: []const u8) void {
    var i: usize = 1;
    while (i < path.len) : (i += 1) {
        if (path[i] != '/') continue;
        var seg: [std.fs.max_path_bytes]u8 = undefined;
        if (i >= seg.len) return;
        @memcpy(seg[0..i], path[0..i]);
        seg[i] = 0;
        _ = std.c.mkdir(seg[0..i :0].ptr, 0o755);
    }
    var full: [std.fs.max_path_bytes]u8 = undefined;
    if (path.len >= full.len) return;
    @memcpy(full[0..path.len], path);
    full[path.len] = 0;
    _ = std.c.mkdir(full[0..path.len :0].ptr, 0o755);
}

/// Write `content` to `dir/name`. Returns false on any failure.
fn writeFile(dir: []const u8, name: []const u8, content: []const u8) bool {
    var pbuf: [std.fs.max_path_bytes]u8 = undefined;
    const path = std.fmt.bufPrintZ(&pbuf, "{s}/{s}", .{ dir, name }) catch return false;
    const fd = std.c.open(path.ptr, .{ .ACCMODE = .WRONLY, .CREAT = true, .TRUNC = true }, @as(c_uint, 0o644));
    if (fd < 0) return false;
    defer _ = std.c.close(fd);
    var off: usize = 0;
    while (off < content.len) {
        const n = std.c.write(fd, content[off..].ptr, content.len - off);
        if (n <= 0) return false;
        off += @intCast(n);
    }
    return true;
}

/// Set up shell integration. Writes the scripts and exports the wiring env
/// vars. When `enabled` is false, exports only the harmless markers and skips
/// `ZDOTDIR`. Any filesystem failure is logged and degrades to "no
/// integration" — never fatal. Call once at startup, before any tab spawns.
pub fn setup(enabled: bool) void {
    var dbuf: [std.fs.max_path_bytes]u8 = undefined;
    const dir = runtimeDir(&dbuf) orelse {
        std.debug.print("caldera-console: shell integration: $HOME unset, skipped\n", .{});
        return;
    };
    mkdirP(dir);

    const ok_zsh = writeFile(dir, "caldera-integration.zsh", integration_zsh);
    const ok_bash = writeFile(dir, "caldera-integration.bash", integration_bash);
    const ok_env = writeFile(dir, ".zshenv", zdotdir_zshenv);
    if (!(ok_zsh and ok_bash and ok_env)) {
        std.debug.print("caldera-console: shell integration: write failed, skipped\n", .{});
        return;
    }

    // Markers — always exported; harmless to any shell.
    _ = setenv("CALDERA_CONSOLE", "1", 1);
    var bbuf: [std.fs.max_path_bytes]u8 = undefined;
    if (std.fmt.bufPrintZ(&bbuf, "{s}/caldera-integration.bash", .{dir})) |bash_path| {
        _ = setenv("CALDERA_SHELL_INTEGRATION", bash_path.ptr, 1);
    } else |_| {}

    if (!enabled) return;

    // zsh auto-injection: point ZDOTDIR at our dir, after stashing the real one.
    const real = std.c.getenv("ZDOTDIR");
    if (real) |r| {
        _ = setenv("CALDERA_REAL_ZDOTDIR", r, 1);
    }
    var zbuf: [std.fs.max_path_bytes]u8 = undefined;
    if (std.fmt.bufPrintZ(&zbuf, "{s}/caldera-integration.zsh", .{dir})) |zsh_path| {
        _ = setenv("CALDERA_SHELL_INTEGRATION_ZSH", zsh_path.ptr, 1);
    } else |_| {}
    var dz: [std.fs.max_path_bytes]u8 = undefined;
    if (std.fmt.bufPrintZ(&dz, "{s}", .{dir})) |dirz| {
        _ = setenv("ZDOTDIR", dirz.ptr, 1);
    } else |_| {}
}

const testing = std.testing;

test "runtimeDir resolves under HOME" {
    // Save and override HOME.
    const saved = std.c.getenv("HOME");
    _ = setenv("HOME", "/tmp/caldera-shell-test", 1);
    defer if (saved) |s| {
        _ = setenv("HOME", s, 1);
    };
    var buf: [std.fs.max_path_bytes]u8 = undefined;
    const dir = runtimeDir(&buf).?;
    try testing.expectEqualStrings("/tmp/caldera-shell-test/.cache/caldera-console/shell", dir);
}

test "setup writes the scripts and exports markers" {
    const saved_home = std.c.getenv("HOME");
    _ = setenv("HOME", "/tmp/caldera-shell-test", 1);
    defer if (saved_home) |s| {
        _ = setenv("HOME", s, 1);
    };

    setup(true);

    // The three files exist and are non-empty.
    var buf: [std.fs.max_path_bytes]u8 = undefined;
    const dir = runtimeDir(&buf).?;
    var pbuf: [std.fs.max_path_bytes]u8 = undefined;
    inline for (.{ "caldera-integration.zsh", "caldera-integration.bash", ".zshenv" }) |name| {
        const path = try std.fmt.bufPrintZ(&pbuf, "{s}/{s}", .{ dir, name });
        const fd = std.c.open(path.ptr, .{ .ACCMODE = .RDONLY }, @as(c_uint, 0));
        try testing.expect(fd >= 0);
        _ = std.c.close(fd);
    }

    // Markers exported.
    try testing.expect(std.c.getenv("CALDERA_CONSOLE") != null);
    try testing.expect(std.c.getenv("ZDOTDIR") != null);
}

test "setup(false) does not export ZDOTDIR" {
    const saved_home = std.c.getenv("HOME");
    _ = setenv("HOME", "/tmp/caldera-shell-test-2", 1);
    defer if (saved_home) |s| {
        _ = setenv("HOME", s, 1);
    };
    // Clear any ZDOTDIR a prior test set.
    _ = unsetenv("ZDOTDIR");

    setup(false);
    try testing.expect(std.c.getenv("CALDERA_CONSOLE") != null); // marker still set
    try testing.expect(std.c.getenv("ZDOTDIR") == null); // but no injection
}

File: crates/anvil-platform/src/shell_integration.rs
//! Shell integration setup: write the embedded zsh/bash scripts to a runtime
//! directory and export the env vars that wire spawned shells to them.
//!
//! Port of `src/app/shell_integration.zig`. The `setup` function must be
//! called once at startup, before any PTY is spawned, so that the env vars
//! are inherited by every shell child process.

use std::env;
use std::fs;
use std::path::PathBuf;

const INTEGRATION_ZSH: &str = include_str!("../../../shell/anvil-integration.zsh");
const INTEGRATION_BASH: &str = include_str!("../../../shell/anvil-integration.bash");
const ZDOTDIR_ZSHENV: &str = include_str!("../../../shell/zdotdir-zshenv.zsh");

/// Absolute path to the `anvil-prompt` binary, resolved next to this
/// executable. Returns `None` if the executable path cannot be determined.
fn prompt_binary_path() -> Option<PathBuf> {
    let exe = env::current_exe().ok()?;
    let dir = exe.parent()?;
    Some(dir.join("anvil-prompt"))
}

/// Resolve `~/.cache/anvil/shell`. Returns `None` when `$HOME` is unset.
fn runtime_dir() -> Option<PathBuf> {
    let home = env::var("HOME").ok()?;
    Some(
        PathBuf::from(home)
            .join(".cache")
            .join("anvil")
            .join("shell"),
    )
}

/// Write `content` to `dir/name`. Returns `false` on any failure (best-effort,
/// never fatal — mirrors the Zig behaviour).
fn write_file(dir: &std::path::Path, name: &str, content: &str) -> bool {
    let path = dir.join(name);
    fs::write(&path, content.as_bytes()).is_ok()
}

/// Set up shell integration. Writes the scripts and exports the wiring env
/// vars. When `enabled` is false, exports only the harmless markers and skips
/// the `ZDOTDIR` injection. Any filesystem failure is logged and degrades to
/// "no integration" — never fatal. Call once at startup, before any tab spawns.
pub fn setup(enabled: bool) {
    let dir = match runtime_dir() {
        Some(d) => d,
        None => {
            eprintln!("anvil: shell integration: $HOME unset, skipped");
            return;
        }
    };

    if fs::create_dir_all(&dir).is_err() {
        eprintln!("anvil: shell integration: mkdir failed, skipped");
        return;
    }

    let ok_zsh = write_file(&dir, "anvil-integration.zsh", INTEGRATION_ZSH);
    let ok_bash = write_file(&dir, "anvil-integration.bash", INTEGRATION_BASH);
    let ok_env = write_file(&dir, ".zshenv", ZDOTDIR_ZSHENV);
    if !(ok_zsh && ok_bash && ok_env) {
        eprintln!("anvil: shell integration: write failed, skipped");
        return;
    }

    // SAFETY: setup() is documented as single-threaded startup only. All
    // set_var/remove_var calls below happen before any PTY threads are spawned,
    // so there is no concurrent env access that could trigger data races.

    // Markers — always exported; harmless to any shell.
    unsafe { env::set_var("ANVIL", "1") };

    let bash_path = dir.join("anvil-integration.bash");
    if let Some(s) = bash_path.to_str() {
        unsafe { env::set_var("ANVIL_SHELL_INTEGRATION", s) };
    }

    if let Some(pp) = prompt_binary_path() {
        if let Some(s) = pp.to_str() {
            unsafe { env::set_var("ANVIL_PROMPT", s) };
        }
    }

    if !enabled {
        return;
    }

    // zsh auto-injection: stash the real ZDOTDIR, then point ZDOTDIR at our
    // runtime dir so our .zshenv shim is sourced first.
    if let Ok(real) = env::var("ZDOTDIR") {
        unsafe { env::set_var("ANVIL_REAL_ZDOTDIR", &real) };
    }

    let zsh_path = dir.join("anvil-integration.zsh");
    if let Some(s) = zsh_path.to_str() {
        unsafe { env::set_var("ANVIL_SHELL_INTEGRATION_ZSH", s) };
    }

    if let Some(s) = dir.to_str() {
        unsafe { env::set_var("ZDOTDIR", s) };
    }
}

// -- tests -------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use std::sync::Mutex;

    use super::*;

    /// Serialise all tests that mutate process env to prevent races.
    static ENV_LOCK: Mutex<()> = Mutex::new(());

    /// Helper: override HOME for the duration of a test, restore on drop.
    struct HomeGuard {
        saved: Option<String>,
    }
    impl HomeGuard {
        fn set(home: &str) -> Self {
            let saved = env::var("HOME").ok();
            // SAFETY: ENV_LOCK is held by the caller, ensuring no concurrent
            // env access from other test threads.
            unsafe { env::set_var("HOME", home) };
            HomeGuard { saved }
        }
    }
    impl Drop for HomeGuard {
        fn drop(&mut self) {
            // SAFETY: ENV_LOCK is held by the caller.
            match &self.saved {
                Some(s) => unsafe { env::set_var("HOME", s) },
                None => unsafe { env::remove_var("HOME") },
            }
        }
    }

    #[test]
    fn runtime_dir_resolves_under_home() {
        let _lock = ENV_LOCK.lock().unwrap();
        let _g = HomeGuard::set("/tmp/anvil-shell-test");
        let dir = runtime_dir().expect("runtime_dir should return Some");
        assert_eq!(
            dir,
            PathBuf::from("/tmp/anvil-shell-test/.cache/anvil/shell")
        );
    }

    #[test]
    fn setup_writes_scripts_and_exports_markers() {
        let _lock = ENV_LOCK.lock().unwrap();
        let _g = HomeGuard::set("/tmp/anvil-rust-shell-test");
        // Remove any stale ZDOTDIR so this test is clean.
        // SAFETY: ENV_LOCK serialises env access across test threads.
        unsafe { env::remove_var("ZDOTDIR") };

        setup(true);

        let dir = runtime_dir().unwrap();
        for name in &["anvil-integration.zsh", "anvil-integration.bash", ".zshenv"] {
            let path = dir.join(name);
            assert!(path.exists(), "{name} should exist");
            let content = fs::read(&path).unwrap();
            assert!(!content.is_empty(), "{name} should be non-empty");
        }

        assert!(env::var("ANVIL").is_ok(), "ANVIL marker should be set");
        assert!(
            env::var("ZDOTDIR").is_ok(),
            "ZDOTDIR should be injected when enabled=true"
        );
        assert!(
            env::var("ANVIL_SHELL_INTEGRATION").is_ok(),
            "ANVIL_SHELL_INTEGRATION should be set"
        );
        assert!(
            env::var("ANVIL_SHELL_INTEGRATION_ZSH").is_ok(),
            "ANVIL_SHELL_INTEGRATION_ZSH should be set"
        );
    }

    #[test]
    fn setup_preserves_a_pre_existing_zdotdir() {
        let _lock = ENV_LOCK.lock().unwrap();
        let _g = HomeGuard::set("/tmp/anvil-rust-zdotdir-test");
        // SAFETY: ENV_LOCK serialises env access across test threads.
        unsafe { env::remove_var("ANVIL_REAL_ZDOTDIR") };
        unsafe { env::set_var("ZDOTDIR", "/tmp/my-zdotdir") };

        setup(true);

        let real = env::var("ANVIL_REAL_ZDOTDIR")
            .expect("ANVIL_REAL_ZDOTDIR should be set when ZDOTDIR was present");
        assert_eq!(real, "/tmp/my-zdotdir");
    }

    #[test]
    fn setup_false_does_not_export_zdotdir() {
        let _lock = ENV_LOCK.lock().unwrap();
        let _g = HomeGuard::set("/tmp/anvil-rust-shell-test-2");
        // SAFETY: ENV_LOCK serialises env access across test threads.
        unsafe { env::remove_var("ZDOTDIR") };

        setup(false);

        assert!(env::var("ANVIL").is_ok(), "ANVIL marker should be set");
        assert!(
            env::var("ZDOTDIR").is_err(),
            "ZDOTDIR must not be injected when enabled=false"
        );
    }
}
